File: example/step-by-step.c
/**
 * @file step-by-step.c
 *
 * @brief A step-by-step example that runs one full analyze (ordering, symbolic
 *       factorization, analyze), then loops over 2 factorizations that are both
 *       used for 2 solves each.
 *
 * @copyright 2015-2023 Bordeaux INP, CNRS (LaBRI UMR 5800), Inria,
 *                      Univ. Bordeaux. All rights reserved.
 *
 * @version 6.3.1
 * @author Mathieu Faverge
 * @author Pierre Ramet
 * @author Tony Delarue
 * @author Alycia Lisito
 * @date 2023-11-07
 *
 * @ingroup pastix_examples
 * @code
 *
 */
#include <pastix.h>
#include <spm.h>

int main (int argc, char **argv)
{
    pastix_data_t  *pastix_data = NULL; /*< Pointer to the storage structure required by pastix */
    pastix_int_t    iparm[IPARM_SIZE];  /*< Integer in/out parameters for pastix                */
    double          dparm[DPARM_SIZE];  /*< Floating in/out parameters for pastix               */
    spm_driver_t    driver;
    char           *filename = NULL;
    spmatrix_t     *spm, spm2;
    void           *x, *b, *x0 = NULL;
    size_t          size;
    int             scatter = 0;
    int             check   = 1;
    int             nrhs    = 10;
    int             rc      = PASTIX_SUCCESS;
    int             nfact   = 2;
    int             nsolv   = 2;
    long            i, j;

    /**
     * Initialize parameters to default values
     */
    pastixInitParam( iparm, dparm );

    /**
     * Get options from command line
     */
    pastixGetOptions( argc, argv,
                      iparm, dparm,
                      &check, &scatter, &driver, &filename );

    /**
     * Startup PaStiX
     */
    pastixInit( &pastix_data, MPI_COMM_WORLD, iparm, dparm );

    /**
     * Read the sparse matrix with the driver
     */
    spm = malloc( sizeof( spmatrix_t ) );
    if ( scatter ) {
        rc = spmReadDriverDist( driver, filename, spm, MPI_COMM_WORLD );
    }
    else {
        rc = spmReadDriver( driver, filename, spm );
    }
    free( filename );
    if ( rc != SPM_SUCCESS ) {
        pastixFinalize( &pastix_data );
        return rc;
    }

    spmPrintInfo( spm, stdout );

    rc = spmCheckAndCorrect( spm, &spm2 );
    if ( rc != 0 ) {
        spmExit( spm );
        *spm = spm2;
        rc = 0;
    }

    /**
     * Generate a Fake values array if needed for the numerical part
     */
    if ( spm->flttype == SpmPattern ) {
        spmGenFakeValues( spm );
    }

    /**
     * Perform ordering, symbolic factorization, and analyze steps
     */
    pastix_subtask_order( pastix_data, spm, NULL );
    pastix_subtask_symbfact( pastix_data );
    pastix_subtask_reordering( pastix_data );
    pastix_subtask_blend( pastix_data );

    /**
     * Normalize A matrix (optional, but recommended for low-rank functionality)
     */
    double normA = spmNorm( SpmFrobeniusNorm, spm );
    spmScal( 1./normA, spm );

    size = pastix_size_of( spm->flttype ) * spm->nexp * nrhs;
    x = malloc( size );
    b = malloc( size );
    if ( check > 1 ) {
        x0 = malloc( size );
    }

    /* Do nfact factorization */
    for (i = 0; i < nfact; i++)
    {
        /**
         * Perform the numerical factorization
         */
        pastix_subtask_spm2bcsc( pastix_data, spm );
        pastix_subtask_bcsc2ctab( pastix_data );
        pastix_subtask_sopalin( pastix_data );
        for (j = 0; j < nsolv; j++)
        {
            /**
             * Generates the b and x vector such that A * x = b
             * Compute the norms of the initial vectors if checking purpose.
             */
            if ( check )
            {
                spmGenRHS( SpmRhsRndX, nrhs, spm, x0, spm->nexp, b, spm->nexp );
                memcpy( x, b, size );
            }
            else {
                spmGenRHS( SpmRhsRndB, nrhs, spm, NULL, spm->nexp, x, spm->nexp );

                /* Apply also normalization to b vectors */
                spmScalMat( 1./normA, spm, nrhs, b, spm->nexp );

                /* Save b for refinement */
                memcpy( b, x, size );
            }

            /**
             * Solve the linear system
             */
            pastix_task_solve( pastix_data, spm->nexp, nrhs, x, spm->nexp );
            pastix_task_refine( pastix_data, spm->nexp, nrhs, b, spm->nexp, x, spm->nexp );

            if ( check ) {
                rc |= spmCheckAxb( dparm[DPARM_EPSILON_REFINEMENT], nrhs, spm, x0, spm->nexp, b, spm->nexp, x, spm->nexp );
            }
        }
    }

    spmExit( spm );
    free( spm );
    free( b );
    free( x );
    if ( x0 ) {
        free( x0 );
    }
    pastixFinalize( &pastix_data );

    return rc;
}

/**
 * @endcode
 */

File: wrappers/julia/examples/step-by-step.jl
#!/usr/bin/env julia
#=
 @file step-by-step.jl

 PaStiX step by step example

 @copyright 2019-2023 Bordeaux INP, CNRS (LaBRI UMR 5800), Inria,
                      Univ. Bordeaux. All rights reserved.

 @version 6.3.0
 @author Mathieu Faverge
 @author Selmane Lebdaoui
 @date 2022-10-11
=#

using Pkg
Pkg.activate("../../spm/wrappers/julia/spm")
Pkg.activate("PaStiX")
using CBinding
using PaStiX
using spm
using Printf

if PaStiX.pastix_mpi_enabled
    using MPI
    MPI.Init()
end

#
# Two solutions to select the outpu file to pass to output functions
#
my_stdout = Libc.FILE( Libc.RawFD(1), "w" ) # Select the stdout or stderr through 1 or 2
#my_stdout = Ptr{Cvoid}(0)                   # Give a null pointer to use the default

# Initialize parameters to default values
iparm = zeros( PaStiX.Pastix_int_t, PaStiX.iparm_size )
dparm = zeros( Cdouble, PaStiX.dparm_size )
PaStiX.pastixInitParam( iparm, dparm )

iparm[PaStiX.iparm_scheduler] = PaStiX.schedsequential # ./simple -s 0

# Startup PaStiX
pastix_data = zeros(Int64,1)
pastix_data_ptr = pointer_from_objref(pastix_data)
PaStiX.pastixInit( pastix_data_ptr, 0, iparm, dparm )

for nsbp in 1:3
    A    = spm.spmatrix_t( zero )
    Aptr = pointer_from_objref( A )

    # Load a sparse matrix
    filename = @sprintf( "%d:%d:%d:4.:1.", nsbp * 5, nsbp * 5, nsbp * 5 )
    spm.spmReadDriver( spm.SpmDriverLaplacian, filename, Aptr )
    spm.spmPrintInfo( Aptr, my_stdout )

    # Scale A for low-rank: A / ||A||_f
    norm = spm.spmNorm( spm.SpmFrobeniusNorm, Aptr )
    spm.spmScal( 1. / norm, Aptr )

    # Perform analyze
    PaStiX.pastix_task_analyze( pastix_data, Aptr )

    for nfact in 1:3
        # Perform numerical factorization
        PaStiX.pastix_task_numfact( pastix_data, Aptr )

        for nsolv in 1:3
            # Generate b and x0 vector such that A * x0 = b
            nrhs = 10
            n    = A.nexp
            X    = zeros(Cdouble, (n, nrhs))
            B    = zeros(Cdouble, (n, nrhs))
            X0   = zeros(Cdouble, (n, nrhs))

            # Generate b and x0 vector such that A * x0 = b
            spm.spmGenRHS( spm.SpmRhsRndX, nrhs, Aptr, X0, n, B, n )

            # Perform solve
            X = copy(B)
            PaStiX.pastix_task_solve( pastix_data, n, nrhs, X, n )

            # Refine the solution
            PaStiX.pastix_task_refine( pastix_data, n, nrhs, B, n, X, n )

            # Check solution
            eps = dparm[PaStiX.dparm_epsilon_refinement]
            spm.spmCheckAxb( eps, nrhs, Aptr, X0, n, B, n, X, n )
        end
    end

    spm.spmExit( Aptr )
end

PaStiX.pastixFinalize( pastix_data_ptr )
